### ISSUE TITLE
Support sanity checking of extension configuration without requiring all dependencies

### ISSUE THREAD

**Issue by akonradi** [envoyproxy/envoy#15032]
I want to deploy my Envoys using my custom FooFilter attached to a static listener. FooFilter has a custom configuration with some fields that must be set, with cross-field dependencies (specifically things that can't be checked using PGV annotations). I want to verify that as I add and remove fields, my checked-in bootstrap configs are still valid.

In a world in which all abstractions are correct, my development environment would perfectly mirror production and there would be no difference in instantiating FooFilter for validation versus execution. This seems like an unreasonable expectation, though, so instead I would like my FooFilter to be able to know somehow whether it should actually expect to find all its dependencies or just sanity check its config to the best of its ability.

Right now, Envoy has a notion of execution versus validation modes, available via `Envoy::Server::Options::mode()`. `Mode::Validate` seems closes to what I want. The implicit contract seems to be that if `Mode::Validate` produces errors, then any other mode will also produce errors (but not the reverse). Extensions have access to information about the current context via the various .*FactoryContext interfaces, but Options is not exposed.

This feature request is two-fold:
1. make `Envoy::Server::Options` available to ResourceMonitorFactoryContext and FilterFactoryContext and optionally other factory contexts
1. clarify the contract between, or at least the documentation of, the various Mode options

**Comment by antoniovicente** [comment]
cc @envoyproxy/api-shepherds 

It seems natural that there will be some components that are available in a production environment but not on an individual developer's workstation.  Possible example: expansion card that provides tamper-proof crypto hardware that holds RSA private key. It seems desirable to have the ability to validate the production config in Mode::Validate on a machine without the specialized hardware.

Exposing this mode to filter and extension constructors seems fine to me. We do need to trust that extension implementations won't look at the mode option more than they have to since each validation step that ends up being slightly different in each mode reduces the fidelity of the config validation.

**Comment by htuch** [comment]
Yeah, this seems a pretty useful thing to add, you might be able to tackle https://github.com/envoyproxy/envoy/issues/9809 with it. I don't think this gets us a strong guarantee that every filter built into Envoy will have high fidelity offline validation, but as @antoniovicente points out, this is a fundamental flaw of the split between config validation and operational mode.

**Comment by akonradi** [comment]
Any objections to adding a `CommonFactoryContext::options()` method [here](https://github.com/envoyproxy/envoy/blob/c676c00bbd371b2ad39eb9e01813a3085812edc4/include/envoy/server/factory_context.h#L43)? And similar to ResourceMonitorFactoryContext? I'm not sure that's the absolute best place but it doesn't cost that much to move it around later. Alternatively, I could see an argument to making it accessible via [LocalInfo](https://github.com/envoyproxy/envoy/blob/c676c00bbd371b2ad39eb9e01813a3085812edc4/include/envoy/local_info/local_info.h#L17).

**Comment by mattklein123** [comment]
In general I am a big +1 on this. I agree with @htuch that it would be great to solve https://github.com/envoyproxy/envoy/issues/9809 also, since it's basically the same issue, even if part of core code.

Also, if you are feeling like extra bonus points, see also https://github.com/envoyproxy/envoy/issues/2636. It would be *really* nice to make the config_test stuff also go through the same path. Right now it's a different path that uses mocks.

**Comment by antoniovicente** [comment]
Sounds like we all think that passing mode options to factories is acceptable.  CommonFactoryContext seems like the source of other similar options like Envoy::Runtime::Loader.  Try it out.

**Comment by akonradi** [comment]
Copying from discussion on #9809 to keep it all in one place:

> In that case, I'd like to propose changing Mode to the following four levels of fidelity:
> 
> - load: load the configuration from disk and make sure it is syntactically valid
> - validate: load the configuration and all required resources from disk, but don't open any network connections
> - initonly: load the configuration and all required resources from disk, do everything except serve traffic
> - serve: load the configuration and all required resources from disk, then serve
